assets/js/modules/analytics/util/index.js

```diff
@@ -45,32 +45,67 @@ export * from './validation';
  * Extracts data required for a pie chart from the Analytics report information.
  *
  * @since 1.16.0 Added keyColumnIndex argument.
+ * @since n.e.x.t Updated the function signature to use options argument instead of keyColumnIndex.
  *
- * @param {Array}  reports        The array with reports data.
- * @param {number} keyColumnIndex The number of a column to extract metrics data from.
+ * @param {Array}    reports                   The array with reports data.
+ * @param {Object}   [options]                 Optional. Data extraction options.
+ * @param {number}   [options.keyColumnIndex]  Optional. The number of a column to extract metrics data from.
+ * @param {boolean}  [options.withOthers]      Optional. Whether to add "Others" record to the data map or not.
+ * @param {Function} [options.tooltipCallback] Optional. A callback function for tooltip column values.
  * @return {Array} Extracted data.
  */
-export function extractAnalyticsDataForTrafficChart( reports, keyColumnIndex ) {
+export function extractAnalyticsDataForPieChart( reports, options = {} ) {
 	if ( ! reports || ! reports.length ) {
 		return null;
 	}
 
+	const {
+		keyColumnIndex = 0,
+		withOthers = false,
+		tooltipCallback,
+	} = options;
+
 	const data = reports[ 0 ].data;
 	const rows = data.rows;
 
+	const withTooltips = typeof tooltipCallback === 'function';
+	const columns = [ 'Source', 'Percent' ];
+	if ( withTooltips ) {
+		columns.push( {
+			type: 'string',
+			role: 'tooltip',
+			p: {
+				html: true,
+			},
+		} );
+	}
+
 	const totalUsers = data.totals[ 0 ].values[ keyColumnIndex ];
-	const dataMap = [
-		[ 'Source', 'Percent' ],
-	];
+	const dataMap = [ columns ];
 
-	each( rows, ( row ) => {
+	let others = 1;
+	for ( const row of rows ) {
 		const users = row.metrics[ 0 ].values[ keyColumnIndex ];
 		const percent = ( users / totalUsers );
 
-		const source = row.dimensions[ 0 ];
+		others -= percent;
 
-		dataMap.push( [ source, percent ] );
-	} );
+		const rowData = [ row.dimensions[ 0 ], percent ];
+		if ( withTooltips ) {
+			rowData.push( tooltipCallback( row, rowData ) );
+		}
+
+		dataMap.push( rowData );
+	}
+
+	if ( withOthers && others > 0.01 ) {
```

Comment:
Based on my previous feedback, here we could rely on whether there are remaining rows instead of checking `others` being greater than 0.01.

Reply:
Updated.